lib/src/clojure_lsp/dep_graph.clj

```diff
@@ -0,0 +1,347 @@
+;; `'aaa`'s :dependencies, and `'aaa` would be in `'bbb`'s :dependents.
+
+;; :aliases is a multiset of the symbols that are used to refer to this
+;; namespace. If the code had `(:require [aaa :as a])`, then `'aaa` would
+;; include `'a` in its aliases. Unaliased requires like `(:require [aaa])` will
+;; include `nil` in their aliases. May be either empty or absent if a namespace
+;; isn't required by other namespaces.
+
+;; :files is a set of the files in which the namespace is defined. This can be
+;; either absent or empty if the namespace is required but never defined, either
+;; as a syntax error or via :as-alias.
+
+;; :internal? is true when the namespace is defined in any internal file, falsy
+;; otherwise. May be absent if the namespace is required but never defined, as
+;; for :files.
+
+;; :from-internal? is true when the namespace is required by any internal
+;; namespace, falsy otherwise. May be absent if the namespace isn't required by
+;; other namespaces.
+
+;; :from-langs is a multiset of the langs of the :namespace-usages or files (if the
+;; :namespace-usages don't have langs) that use this namespace. May be either
+;; empty or absent if the namespace isn't required by other namespaces.
+
+;; file-meta-item =
+;; {(:namespaces #{ns*})?
+;;  :internal? boolean
+;;  :langs #{lang+}}
+
+;; :namespaces is a set of the namespaces defined by this file. May be either
+;; empty or absent if the file doesn't define any namespaces.
+
+;; :internal? is whether this file is internal to the project.
+
+;; :langs is a set of the langs used by this file.
+
+;; file-meta = {(filename file-meta-item)*}
+;; dep-graph = {(ns dep-graph-item)*}
+;; db = {':dep-graph' dep-graph
+;;       ':file-meta' file-meta}
+
+(comment
+  ;; Example
+  (require '[clojure-lsp.db :as db])
+  (-> @db/db*
+      :dep-graph
+      (select-keys '[clojure-lsp.main
+                     clojure.tools.cli]))
+  '{clojure-lsp.main  {:dependents     {clojure-lsp.main-test 1}
+                       :aliases        {main 1}
+                       :from-internal? true
+                       :from-langs     {:clj 1}
+                       :dependencies   {borkdude.dynaload        1
+                                        clojure-lsp.internal-api 1
+                                        clojure-lsp.kondo        1
+                                        clojure-lsp.server       1
+                                        clojure-lsp.shared       1
+                                        clojure.core             1
+                                        clojure.edn              1
+                                        clojure.java.io          1
+                                        clojure.string           1
+                                        clojure.tools.cli        1
+                                        pod.clojure-lsp.api      1}
+                       :files          #{"~/code/clojure-lsp/cli/src/clojure_lsp/main.clj"}
```

Comment:
Usually, I call `file` a java.io.file and let filename for strings, it'd follow the rest of the codebase naming I think too

Reply:
One thing to keep in mind here is to maybe use always the absolute path instead of relative and in the future think about persists as URI instead of filename, there is an idea of kondo use URI for everything too for better compatibility with other features like remote LSP

Reply:
> filename for strings

I thought about that a couple of times as I was working on this—it's a good point. I'll refactor. 

> the absolute path

In the DB it's an absolute path. I replaced `/Users/jmaine` with `~` so that my name wasn't all over the example, but I'll find another way so that it isn't ambiguous. Maybe `/Users/me`.

I thought about using URI as a key. I'd rather do that too. But the dep graph is mostly about speeding up access to kondo data, so for now we need file names. Maybe those things should be changed in tandem.

Reply:
I realized that later 😅 , I'm ok with something simple like `/path/to/clojure-lsp-project/...` or even the `/Users/me`, feel free to choose

Reply:
Let me think about using URIs a bit more...

URI pros:
- It'd be nice to avoid refactoring the dep-graph when kondo starts returning URIs.

neutral:
- When kondo starts returning URIs, we'll need a db/version bump, anyway.

URI cons:
- To use the dep-graph to access analysis, we'd need an extra pass to convert URIs to filenames. We'd cache filename by URI, limiting the expense. Still, needs performance testing.
- We won't have a way to cache the other direction, filename to URI. We've had performance problems when we accidentally do this conversion in a loop. Need to audit code, to see if we often need to convert a bunch of filenames to URIs.

Reply:
I moved the file metadata into `:documents`. Looks fine from a performance perspective.

Thanks for the suggestion... I think it's nicer this way. It'll make it easier to refactor to URIs later.

To keep things fast, we still have a `:file-meta` key in the db, but all it does is map filename to uri. We'll be able to remove it when we use URIs consistently.